### PR TITLE
ISSUE #945: Enabling BookieShell for LocalBookKeeper

### PR DESCRIPTION
**Pull request by reddycharan**
Descriptions of the changes in this PR:

We should be able to execute BookieShell commands against bookies of LocalBookkeeper

eg. for Zookeeper/Metadata Operation
./bookkeeper shell -localbookie listbookies -rw

eg. for Bookie operation
./bookkeeper shell -localbookie 10.3.27.190:5000 lastmark
./bookkeeper shell -localbookie 10.3.27.190:5000 listfilesondisc

Master Issue: #945 

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
>
> If this PR is a BookKeeper Proposal (BP):
>
> - [ ] Make sure the PR title is formatted like:
>     `<BP-#>: Description of bookkeeper proposal`
>     `e.g. BP-1: 64 bits ledger is support`
> - [ ] Attach the master issue link in the description of this PR.
> - [ ] Attach the google doc link if the BP is written in Google Doc.
>
> Otherwise:
> 
> - [ ] Make sure the PR title is formatted like:
>     `<Issue # or BOOKKEEPER-#>: Description of pull request`
>     `e.g. Issue 123: Description ...`
>     `e.g. BOOKKEEPER-1234: Description ...`
> - [ ] Make sure tests pass via `mvn clean apache-rat:check install spotbugs:check`.
> - [ ] Replace `<Issue # or BOOKKEEPER-#>` in the title with the actual Issue/JIRA number.
> 
> ---

### CODE REVIEW

File: bookkeeper-server/conf/bk_server.conf [hunk 1]
```diff
@@ -48,6 +48,12 @@ journalDirectory=/tmp/bk-txn
 # Directories to store index files. If not specified, will use ledgerDirectories to store.
 # indexDirectories=/tmp/bk-data
 
+# Directory where LocalBookKeeper serializes the bookies config objects, which can be 
+# later used by BookieShell for running against LocalBookKeeper.
+# This config value should match DEFAULT_LOCALBOOKIES_CONFIG_DIR value in  
+# bookkeeper/bookkeeper-server/bin/bookkeeper script file.
+# localBookiesConfigDirectory=/tmp/localbookies-config
```
Comment (eolivelli) [inline bookkeeper-server/conf/bk_server.conf]: This is used only for testing and devs maybe it is better to move to the very bottom of the file?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/util/LocalBookKeeper.java [hunk 2]
```diff
@@ -224,13 +237,35 @@ private void runBookies(ServerConfiguration baseConf, List<File> tempDirs, Strin
...
+         * having correct zk details and this could be used for running any such
+         * bookieshell commands if bookieid is not provided as parameter to
+         * bookkeeper shell operation. for eg:
+         * "./bookkeeper shell localbookie listbookies -rw". But for execution
+         * shell command of bookie Operation nature we need to provide bookieid,
+         * for eg "./bookkeeper shell -localbookie 10.3.27.190:5000 lastmark",
+         * so this shell command would use '10.3.27.190:5000.conf' file
+         */
+        ServerConfiguration baseConfWithCorrectZKServers = new ServerConfiguration(
+                (ServerConfiguration) baseConf.clone());
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/util/LocalBookKeeper.java]: Maybe clone is not needed

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/util/LocalBookKeeper.java [hunk 3]
```diff
@@ -316,6 +351,43 @@ static void startLocalBookiesInternal(ServerConfiguration conf,
...
+        File localBookiesConfigDir = new File(localBookieConfig.getLocalBookiesConfigDirName());
+        File localBookieConfFile = new File(localBookiesConfigDir, fileName);
+        if (localBookieConfFile.exists() && !localBookieConfFile.delete()) {
+            throw new IOException(
+                    "Unable to delete the existing LocalBookieConfigFile - " + localBookieConfFile.getAbsolutePath());
+        }
+        if (!localBookieConfFile.createNewFile()) {
+            throw new IOException("Unable to create new File - " + localBookieConfFile.getAbsolutePath());
+        }
+        PrintWriter writer = new PrintWriter(localBookieConfFile);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/util/LocalBookKeeper.java]: You are not setting file encoding, better to explicitly use utf-8.

Nit: use try-with-resources block

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java [hunk 4]
```diff
@@ -770,6 +771,26 @@ public ServerConfiguration setAllowStorageExpansion(boolean val) {
         return this;
     }
 
+    /**
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java]: I am not sure we should put this settings in a bookie config. A server configuration is specific to a bookie instance.

A better solution is:  LocalBookKeeper takes parameter from command line args. e.g. LocalBookKeeper <numBookies> <localbookie config>. so this setting is only for LocalBookKeeper.

File: bookkeeper-server/conf/bk_server.conf (hunk 1 above)
Comment (reddycharan) [reply bookkeeper-server/conf/bk_server.conf]: going to remove this config

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/util/LocalBookKeeper.java (hunk 3 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/util/LocalBookKeeper.java]: fixed it

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java (hunk 4 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java]: removed config and moved it to localbookkeeper args

**Comment by sijie** [comment]
@eolivelli please review this again to unblock @reddycharan

**Comment by reddycharan** [comment]
@sijie @eolivelli Thanks for the review. It is good to go, right?

**Comment by sijie** [comment]
yes it is good to go. feel free to merge it (or I can do it later)

**Comment by sijie** [comment]
@reddycharan merged this PR. thank you for this great contribution!

**Pull request closed without merging**